### PR TITLE
Add gradient for tensor product basis

### PR DESCRIPTION
**Pull request by alexfikl**

### CODE REVIEW

File: modepy/tools.py [hunk 1]
```diff
@@ -397,6 +400,42 @@ def try_add_tet(d1, d2, d3, d4):
 submesh = MovedFunctionDeprecationWrapper(simplex_submesh)
 
 
+def hypercube_submesh(node_tuples):
```
Comment (alexfikl) [inline modepy/tools.py:403]: Shamelessly copied from [meshmode](https://github.com/inducer/meshmode/blob/master/meshmode/mesh/tools.py#L54).

**Review by inducer — APPROVED** [review]
Thanks! LGTM, just one question. :)

File: modepy/modes.py [hunk 2]
```diff
@@ -648,6 +662,41 @@ def tensor_product_basis(dims, basis_1d):
...
+    func = (basis_1d, grad_basis_1d)
+    return tuple(
+            _TensorProductGradientBasisFunction(order, [
+                [func[i][k] for i, k in zip(iderivative, order)]
+                for iderivative in wandering_element(dims)
+                ])
+            for order in gnitb(len(basis_1d), dims))
+
+
+def legendre_tensor_product_basis(dims, order):
```
Comment (alexfikl) [inline modepy/modes.py:688]: Should these guys be exported in `modepy.__init__`? I mostly added them for testing to match the interface of `simplex_onb`.

File: modepy/tools.py (hunk 1 above)
Comment (alexfikl) [reply modepy/tools.py:403]: Started bumping versions in eaa8500.

What do you think of the rename? Is it ok?

File: modepy/modes.py (hunk 2 above)
Comment (inducer) [reply modepy/modes.py:689]: I think they're common enough that that makes sense.

File: modepy/tools.py (hunk 1 above)
Comment (inducer) [reply modepy/tools.py:403]: Rename is fine.

File: modepy/modes.py [hunk 3]
```diff
@@ -648,6 +663,41 @@ def tensor_product_basis(dims, basis_1d):
...
 
+
+def grad_tensor_product_basis(dims, basis_1d, grad_basis_1d):
+    """Provides derivatives for each of the basis functions generated by
+    :func:`tensor_product_basis`.
+
+    :returns: a :class:`tuple` of callables, where each one returns a
+        *dims*-dimensional :class:`tuple`, one for each derivative.
+
+    .. versionadded: 2020.2
```
Comment (inducer) [inline modepy/modes.py:674]:
```suggestion
    .. versionadded:: 2020.2
```
Does it render properly with just one colon?

Comment (alexfikl) [reply modepy/modes.py:674]: Nope, it was just a typo!

File: modepy/modes.py (hunk 2 above)
Comment (alexfikl) [reply modepy/modes.py:689]: Done in 47a416f.

**Comment by inducer** [comment]
Thanks for working on this!

**Pull request merged**